### PR TITLE
Interface based API for lists

### PR DESCRIPTION
**Pull request by rorbech**
This moves implementation details of `RealmList` into `internal` and changes internal implementation to use `ManagedRealmList` to avoid having to handle unmanaged lists in the Realm-faced implementation.

### CODE REVIEW

File: packages/library/src/commonMain/kotlin/io/realm/RealmList.kt [hunk 1]
```diff
@@ -16,232 +16,31 @@
...
-            }
-        }
-    }
+interface RealmList<E> : MutableList<E> {
+    fun observe(): Flow<RealmList<E>>
 }
 
 /**
- * Delegate interface for Realm-specific methods.
+ * Instantiates a [RealmList] in **unmanaged** mode.
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmList.kt:43]:
```suggestion
 * Instantiates an **unmanaged** [RealmList].
```

Talking about a "mode" make it seems like you can easily switch it IMO.

File: packages/library/src/commonMain/kotlin/io/realm/RealmList.kt [hunk 2]
```diff
@@ -16,232 +16,31 @@
...
-        return RealmList(liveList, metadata.copy(realm = liveRealm))
-    }
-
-    private fun rangeCheckForAdd(index: Int) {
-        if (index < 0 || index > size) {
-            throw IndexOutOfBoundsException("Index: '$index', Size: '$size'")
-        }
-    }
-}
+fun <T> realmListOf(vararg elements: T): RealmList<T> =
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmList.kt:45]: Should we also have an override that takes other collections?

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt [hunk 3]
```diff
@@ -240,19 +240,30 @@ internal class SuspendableNotifier(
...
+            liveComponentPointer = liveListPointer,
             notifyComponentUpdate = { frozenRealm ->
-                if (!liveList.isValid()) {
+                if (!liveListDelegate.isValid()) {
                     callback.onChange(null)
                 } else {
                     // Notifications need to be delivered with the version they where created on, otherwise
                     // the fine-grained notification data might be out of sync.
-                    val frozenList = liveList.freeze(frozenRealm)
+                    val frozenListPointer = RealmInterop.realm_list_freeze(
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:262]: Shouldn't this logic be inside the `liveListDelegate.freeze` method?

File: packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt [hunk 4]
```diff
@@ -377,17 +377,17 @@ class AccessorModifierIrGeneration(private val pluginContext: IrPluginContext) {
...
-                                // This is due to Kotlin Native being unable to retrieve the related
-                                // supertypes in runtime
-                                if (collectionType != CollectionType.NONE) {
-                                    val supertypes = (type as IrSimpleType).classifier.descriptor
-                                        .typeConstructor.supertypes
-                                    if (inheritsFromRealmObject(supertypes)) {
-                                        putValueArgument(2, irBoolean(true))
-                                    }
-                                }
+//                            // Check if the collection type is a subclass of RealmObject
```
Comment (cmelchior) [inline packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt:380]: Why are this commented out?

File: packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt (hunk 3 above)
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt:262]: Yep. Just migrated a previous sample implementation, so didn't gave it too much thought.

File: packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt (hunk 4 above)
Comment (rorbech) [reply packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt]: According to @edualonso (who did the initial prototype of this PR) we don't longer need this parameter as the `RealmObject` case is now handled as a fallback if the argument is not a primitive type.

File: packages/library/src/commonMain/kotlin/io/realm/RealmList.kt (hunk 2 above)
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/RealmList.kt:45]: I added a `Iterable<T>.toRealmList()` extension method following the conventions from Kotlin.

**Review by edualonso — APPROVED** [review]
🚀

**Pull request merged**